File: src/cpp/mayer/pair_potential.cpp
#include "pair_potential.h"

#include <cmath>
#include <stdexcept>

double PairPotential::prob_fij(double rij) const {
    return std::abs(fij(rij));
}

double PairPotential::compute_integrand(const double* pos, int virial_no) const {
    static constexpr double origin[3] = {0.0, 0.0, 0.0};
    const double* p1 = pos;
    const double* p2 = pos + 3;
    const double* p3 = pos + 6;

    if (virial_no == 2) {
        double r12 = dist(origin, p1);
        return fij(r12);
    } else if (virial_no == 3) {
        double r12 = dist(origin, p1);
        double r13 = dist(origin, p2);
        double r23 = dist(p1, p2);
        return fij(r12) * fij(r13) * fij(r23);
    } else if (virial_no == 4) {
        double r12 = dist(origin, p1);
        double r13 = dist(origin, p2);
        double r23 = dist(p1, p2);
        double r34 = dist(p2, p3);
        double r14 = dist(origin, p3);
        double r24 = dist(p1, p3);
        double gamma = fij(r12) * fij(r23) * fij(r34) * fij(r14);
        gamma *= (3.0 + 6.0 * fij(r13) +  fij(r24) * fij(r13));
        return gamma;
    }
    throw std::invalid_argument("Virial coefficient not implemented");
}

double PairPotential::dist(const double* a, const double* b) {
    double dx = a[0] - b[0];
    double dy = a[1] - b[1];
    double dz = a[2] - b[2];
    return std::sqrt(dx * dx + dy * dy + dz * dz);
}

File: src/msmc/mayer/pair_potential.py
import numpy as np

import msmc._core as _core


class PairPotential(_core.PairPotential):
    """Base class for pairwise-additive potential models."""

    def fij(self, rij):
        """
        Mayer-f function for a pair of molecules.

        Parameters
        ----------
        rij : float
            Separation distance between molecule i and j.

        Returns
        -------
        float
            Mayer-f function value.
        """
        return super().fij(rij)

    def prob_fij(self, rij):
        """
        Sampling probability for a pair of molecules.

        Parameters
        ----------
        rij : float
            Separation distance between molecule i and j.

        Returns
        -------
        float
            Absolute value of Mayer-f function.
        """
        return super().prob_fij(rij)

    def compute_integrand(self, pos, virial_no):
        """
        Compute cluster integral integrand for virial coefficients B2-B4.

        Cluster integral corresponds to Mayer-f function for
        pairwise-additive models (Benjamin et al, 2007).

        Parameters
        ----------
        pos : np.ndarray, shape (N, 3)
            Position vectors of particles relative to particle 1 at origin.
        virial_no : int
            Virial coefficient number. 2 <= virial_no <= 4.

        Returns
        -------
        float
            Integrand value of cluster integral.
        """
        return super().compute_integrand(
            np.ascontiguousarray(pos, dtype=np.float64), virial_no
        )
